### PR TITLE
Turn EEx.SyntaxError into a warning to keep backwards compatibility

### LINKED ISSUE #6216: EEx throws SyntaxError on `<%= end %>` on Elixir v1.5-dev
### Environment

* Elixir & Erlang versions: Elixir 1.5.0-dev, OTP 19.3
* Operating system: macOS Sierra 10.12.5

### Current behavior

On Elixir v1.5-dev:

```elixir
== Compilation error in file web/views/some_view.ex ==
** (EEx.SyntaxError) web/templates/some_template.html.eex:57: unexpected beginning of EEx tag "<%=" on end of expression "<%= end %>", please remove "=" accordingly
    (eex) lib/eex/compiler.ex:75: EEx.Compiler.generate_buffer/4
    (eex) lib/eex/compiler.ex:45: EEx.Compiler.generate_buffer/4
    (phoenix) lib/phoenix/template.ex:378: Phoenix.Template.compile/2
    (phoenix) lib/phoenix/template.ex:186: anonymous fn/3 in Phoenix.Template."MACRO-__before_compile__"/2
    (elixir) lib/enum.ex:1812: Enum."-reduce/3-lists^foldl/2-0-"/3
    (phoenix) expanding macro: Phoenix.Template.__before_compile__/1
    web/views/some_view.ex:1: SomeView (module)
```

It's a breaking change, on Elixir <= 1.4 it does not raise.

### Expected behavior

Print warning (similar to `<%= else %>` on master)

```
warning: unexpected beginning of EEx tag "<%=" on "<%= end %>", please remove "=" accordingly
  web/templates/some_template.html.eex:45
```

### PR DESCRIPTION
**Pull request by wojtekmach**
Closes #6216

### CODE REVIEW

File: lib/eex/test/eex_test.exs [hunk 1]
```diff
@@ -175,7 +175,7 @@ defmodule EExTest do
     end
 
     test "when end expression has a modifier" do
-      assert_raise EEx.SyntaxError, ~s[nofile:1: unexpected beginning of EEx tag "<%=" on end of expression "<%= end %>", please remove "=" accordingly], fn ->
+      assert ExUnit.CaptureIO.capture_io :stderr, fn ->
```
Comment (ericmj) [inline lib/eex/test/eex_test.exs]: This no longer tests that the error/warning is generated.

Comment (wojtekmach) [inline lib/eex/test/eex_test.exs]: I initially had it as:

```elixir
      assert ExUnit.CaptureIO.capture_io(:stderr, fn ->
        EEx.compile_string "foo <%= if true do %>true<% else %>false<%= end %>"
      end) =~ ~s[beginning of EEx tag \"<%=\" on end of expression \"<%= end %>\", please remove \"=\" accordingly
```

but noticed that the test above (with warning also) does not check the message. Is asserting on message preferred here and there?

Comment (ericmj) [inline lib/eex/test/eex_test.exs]: Not sure what you mean. The old test asserted on the exception message and this test needs to assert on the warning message. What do you mean that the message is not checked?

Comment (wojtekmach) [inline lib/eex/test/eex_test.exs]: sorry, I was referring to an already existing _different_ test which that did not check the exception message: https://github.com/wojtekmach/elixir/blob/36b9353aa4a0ebbc1192adf607d3d5e2934b67dd/lib/eex/test/eex_test.exs#L171:L175. Thought it was on purpose or something.

anyhow, will push assertions shortly.

File: lib/eex/lib/eex/compiler.ex [hunk 2]
```diff
@@ -69,10 +69,13 @@ defmodule EEx.Compiler do
     {buffer, rest}
   end
 
-  defp generate_buffer([{:end_expr, line, modifier, chars} | _], _buffer, [_ | _], state) do
+  defp generate_buffer([{:end_expr, line, modifier, chars} | t], buffer, [_ | _] = scope, state) do
     message = "unexpected beginning of EEx tag \"<%#{modifier}\" on end of expression \"<%#{modifier}#{chars}%>\", " <>
               "please remove \"#{modifier}\" accordingly"
-    raise EEx.SyntaxError, message: message, file: state.file, line: line
+    :elixir_errors.warn line, state.file, message
```
Comment (whatyouhide) [inline lib/eex/lib/eex/compiler.ex]: I think we are supposed to use `IO.warn` here? 🤔

File: lib/eex/lib/eex/compiler.ex [hunk 3]
```diff
@@ -69,10 +69,13 @@ defmodule EEx.Compiler do
...
+    generate_buffer([{:end_expr, line, '', chars} | t], buffer, scope, state)
+    # TODO: Make this an error on Elixir v2.0 since it accidentally worked previously.
```
Comment (whatyouhide) [inline lib/eex/lib/eex/compiler.ex]: I would move this `TODO` before the actual warning and remove the `raise ...` line. :)

File: lib/eex/lib/eex/compiler.ex (hunk 2 above)
Comment (wojtekmach) [inline lib/eex/lib/eex/compiler.ex]: I don't see an option to omit mod/fun/arity from `IO.warn` call and so I end up with this:

```elixir
IO.warn message, [{:foo, :bar, 0, [file: state.file, line: line]}]

warning: unexpected beginning of EEx tag "<%=" on "<%= else %>", please remove "=" accordingly
  nofile:1: :foo.bar/0
```

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**